### DISCUSSION TITLE
Scan Devices (for educational purpose)

### DISCUSSION THREAD

**Discussion by secmonteiro** [community/community#136723] [Code Security] [answered]
### Select Topic Area

Question

### Body

How can i Develop a scan of devices connected to my network in python?

**Comment by 4po** [comment, marked as answer]
Hello, @FelipeCeub

Developing a network scanner in Python to detect devices connected to your network can be an interesting educational project! Here’s a basic outline to get you started:

### 1. Using Python Libraries

- You can use libraries like socket, subprocess, and scapy to create a network scanner. Here’s a simple example using scapy:

```python
from scapy.all import ARP, Ether, srp

def scan_network(ip_range):
    # Create an ARP request packet
    arp = ARP(pdst=ip_range)
    ether = Ether(dst="ff:ff:ff:ff:ff:ff")
    packet = ether/arp

    # Send the packet and capture the response
    result = srp(packet, timeout=3, verbose=0)[0]

    # Parse the response
    devices = []
    for sent, received in result:
        devices.append({'ip': received.psrc, 'mac': received.hwsrc})

    return devices

# Replace '192.168.1.1/24' with your network's IP range
ip_range = "192.168.1.1/24"
devices = scan_network(ip_range)

print("Devices found:")
for device in devices:
    print(f"IP: {device['ip']}, MAC: {device['mac']}")

```
This script sends an ARP request to the specified IP range and listens for responses from devices on the network.

### 2. Understanding the Code
- ARP (Address Resolution Protocol) is used to map IP addresses to MAC addresses on a local network.
- Ether and ARP are combined to create the packet that is sent across the network.
- **srp** sends the packet and receives the response, allowing you to identify active devices.

### 3. Running the Script
- Make sure to run the script with appropriate permissions (e.g., as an administrator or with sudo on Linux) because network scanning may require elevated privileges.
- Always ensure you have permission to scan a network—unauthorized scanning can be illegal or against network policies.

### 4. Learning Resources
- If you're new to networking or Python scripting, I recommend looking into resources like "Python for Network Engineers" and "Automate the Boring Stuff with Python." These can help you understand the underlying concepts better.

### 5. Caution
-  While this is a great learning project, be mindful of ethical considerations and legal implications. Only scan networks where you have explicit permission to do so.

> Good luck with your project! Feel free to ask if you have any more questions.

4po

**Reply by secmonteiro** [reply]
thanks man, i appreciate you.
